util/workgraph/workgraph.go

```diff
@@ -0,0 +1,449 @@
+// their dependencies have been satisfied, and returning any errors that occur.
+func (g *WorkGraph) Run(ctx context.Context) error {
+	groups, err := g.topoSortKahn()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	type queueEntry struct {
+		id   string
+		done chan struct{}
+	}
+
+	// Create one goroutine that pushes jobs onto our queue...
+	var wg sync.WaitGroup
+	queue := make(chan queueEntry)
+	publishCtx, publishCancel := context.WithCancel(ctx)
+	defer publishCancel()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(queue)
+
+		// For each parallel group...
+		var dones []chan struct{}
+		for _, group := range groups {
+			dones = dones[:0] // re-use existing storage, if any
+
+			// Push all items in this group onto our queue
+			for curr := range group {
+				done := make(chan struct{})
+				dones = append(dones, done)
+
+				select {
+				case <-publishCtx.Done():
+					return
+				case queue <- queueEntry{curr, done}:
+				}
+			}
+
+			// Now that we've started everything, wait for them all
+			// to complete.
+			for _, done := range dones {
+				select {
+				case <-publishCtx.Done():
+					return
+				case <-done:
+				}
+			}
+
+			// Now that we've done this entire group, we can
+			// continue with the next one.
+		}
+	}()
+
+	firstErr := make(chan error, 1)
+	saveErr := func(err error) {
+		if err == nil {
+			return
+		}
+
+		// Tell the publisher to shut down
+		publishCancel()
+
+		select {
+		case firstErr <- err:
+		default:
+		}
+	}
+
+	// ... and N goroutines that each work on an item from the queue.
+	n := g.Concurrency
+	if n == 0 {
+		n = runtime.GOMAXPROCS(-1)
+	}
+
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case ent, ok := <-queue:
+					if !ok {
+						return
+					}
+
+					// immediately call a function so we
+					// can handle panics
+					err := func() (retErr error) {
+						defer close(ent.done)
+						defer func() {
+							if r := recover(); r != nil {
+								// Ensure that we wrap an existing error with %w so errors.Is works
+								switch v := r.(type) {
+								case error:
+									retErr = fmt.Errorf("node %q: caught panic: %w", ent.id, v)
+								default:
+									retErr = fmt.Errorf("node %q: caught panic: %v", ent.id, v)
+								}
+							}
+						}()
+						node := g.nodes[ent.id]
+						return node.Run(ctx)
+					}()
+					if err != nil {
+						saveErr(err)
+						return
+					}
+				}
+			}
+		}()
+	}
+
+	wg.Wait()
+	select {
+	case err := <-firstErr:
+		return err
+	default:
+	}
+	return nil
+}
+
+// Depth-first toplogical sort; used in tests
+//
+// https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
+func (g *WorkGraph) topoSortDFS() (sorted []string, err error) {
+	temporary := make(map[string]bool)
+	permanent := make(map[string]bool)
```

Comment:
could use a single map with an int 1 == temp, 2 == final

Reply:
That seems to have worked, yeah. Also added a fuzz test for the DFS-based topological sort, to be safe.